File: problems/daily/Java/2958. Length of Longest Subarray With at Most K Frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer,Integer> tent = new HashMap<>();
        int ans = 0;
        int l =0;
        for(int i = 0; i < nums.length; i++){
            tent.put(nums[i],tent.getOrDefault(nums[i],0) + 1);
            while(tent.get(nums[i]) > k){
                tent.put(nums[l],tent.get(nums[l]) - 1);
                l++;
            }
            ans = Math.max(ans,i-l+1);
        }
        return ans;
    }
}

File: problems/daily/Python3/2958. Length of Longest Subarray With at Most K Frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        #尺取り法 (使用条件：ウィンドウが大きくなるほど条件が悪化する方向にしか動かない、単調性)
        left = 0
        count = defaultdict(int)
        ans = 0

        for right in range(len(nums)):
            count[nums[right]] += 1
            while count[nums[right]] > k:
                count[nums[left]] -= 1
                left += 1
            ans = max(ans, right - left + 1)

        return ans